{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "23e571f6",
   "metadata": {},
   "source": [
    "## Process amplicon sequences (16S + ITS)\n",
    "\n",
    "Original from Christian Santos Medellin, rewritten by Anneliek ter Horst, May 2023\n",
    "- Take amplicon libraries\n",
    "- Demultiplex\n",
    "- Check quality of libraries\n",
    "- Run clustering algorithm, generating final OTU table and taxonomy table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc49a5cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a file structure\n",
    "# I have 2 lanes, 16S and ITS\n",
    "\n",
    "# go to file folder\n",
    "cd /home/amhorst/amplicon_seq/ \n",
    "\n",
    "# Make needed directories\n",
    "mkdir reads dada_files qual_plots scripts err log\n",
    "cd reads\n",
    "mkdir raw_16s raw_its \n",
    "mkdir raw_16s/FWD raw_16s/RVS raw_16s/filtered\n",
    "mkdir raw_its/FWD raw_its/RVS raw_its/filtered\n",
    "\n",
    "# Place reads in appropriate folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7d45fe4",
   "metadata": {},
   "source": [
    "# Demultiplex reads\n",
    "- Use bananastand\n",
    "- https://github.com/bulksoil/BananaStand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05acf7aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Demultiplex reads\n",
    "cd /home/amhorst/amplicon_seq/scripts ## change this to your corresponding directory\n",
    "sbatch --output=../log/dmx1.log --error=../err/dmx1.err dmx1.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95388d8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dmx.sh\n",
    "\n",
    "# loading conda env\n",
    "source /home/csantosm/initconda\n",
    "conda activate PYTHON2.7\n",
    "\n",
    "# running commands\n",
    "cd /home/amhorst/amplicon_seq/reads/raw_16s \n",
    "python /home/csantosm/BananaStand/demultiplex.py \\\n",
    "-f ./Undetermined_S0_L001_R1_001.fastq.gz \\\n",
    "-r ./Undetermined_S0_L001_R2_001.fastq.gz \\\n",
    "--I1 ./Undetermined_S0_L001_I1_001.fastq.gz \\\n",
    "--I2 ./Undetermined_S0_L001_I2_001.fastq.gz \\\n",
    "-m dmx1.map \\\n",
    "-p T16S \\\n",
    "-a \"dada2\"\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d40e6383",
   "metadata": {},
   "source": [
    "## Check read quality\n",
    "- Plot the quality of libraries using dada2 in R\n",
    "- Plotting scripts from Christian\n",
    "- How to inspect read quality profiles: https://benjjneb.github.io/dada2/tutorial.html\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb7d1e14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# qual plot code\n",
    "# Load libraries\n",
    "library(Rcpp, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(farver, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(labeling, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(digest, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(dada2, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "\n",
    "# Setting the absolute paths for FWD and RVS files\n",
    "fwd1.path <- \"/home/amhorst/amplicon_seq/reads/raw_16s/FWD/\" # Forward\n",
    "rvs1.path <- \"/home/amhorst/amplicon_seq/reads/raw_16s/RVS/\" # reverse\n",
    "\n",
    "# Set prefixes for the files\n",
    "fnFs1 <- sort(list.files(fwd1.path, pattern=\"T16S.*fastq\", full.names = TRUE)) \n",
    "fnRs1 <- sort(list.files(rvs1.path, pattern=\"T16S.*fastq\", full.names = TRUE)) \n",
    "\n",
    "# Plot first 10 forward libraries\n",
    "pdf(\"/home/amhorst/amplicon_seq/qual_plots/raw_T16S_FWD.pdf\")\n",
    "plotQualityProfile(fnFs1[1:10])\n",
    "dev.off()\n",
    "\n",
    "# Plot first 10 reverse libraries\n",
    "pdf(\"/home/amhorst/amplicon_seq/qual_plots/raw_T16S_RVS.pdf\")\n",
    "plotQualityProfile(fnRs1[1:10])\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3541919",
   "metadata": {},
   "source": [
    "## Run the clustering algorithm\n",
    "- Using dada2 in R\n",
    "- For taxonomy of 16S: Use silva_v138.1 database\n",
    "- For taxonomy of ITS: Use the UNITE database (v 10.05.2021)\n",
    "\n",
    "Output files:\n",
    "\n",
    "- Plots for the error training models (stored in the qual_plots folder). Read this if you want to learn more: https://benjjneb.github.io/dada2/tutorial.html\n",
    "- otu1.RDS (stored in the dada_files folder) - formatted OTU table ready to use in R\n",
    "- taxa1.RDS (stored in the dada_files folder) - formatted taxa file ready to use in R\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "42e783f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load libraries\n",
    "library(Rcpp, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(stringr, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(farver, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(labeling, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(digest, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(magrittr, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(dplyr, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(tidyr, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "library(dada2, lib.loc=\"/home/csantosm/R_Packages/R3.6.3/\")\n",
    "\n",
    "# Setting the absolute paths for FWD and RVS files\n",
    "fwd.path <- \"/home/amhorst/amplicon_seq/reads/raw_16s/FWD/\"  ### change this to your path with the demultiplexed forward reads for this lane (lane 1 in this case)\n",
    "rvs.path <- \"/home/amhorst/amplicon_seq/reads/raw_16s/RVS/\"  ### change this to your path with the demultiplexed forward reads for this lane (lane 1 in this case)\n",
    "\n",
    "# Forward and reverse fastq filenames have format: T16S.fastq\n",
    "fnFs <- sort(list.files(fwd.path, pattern=\"T16S.*fastq\", full.names = TRUE)) ### change the pattern to fit the prefix used for your SampleIDs, in this case the prefix is \"WUF\"\n",
    "fnRs <- sort(list.files(rvs.path, pattern=\"T16S.*fastq\", full.names = TRUE)) ### change the pattern to fit the prefix used for your SampleIDs, in this case the prefix is \"WUF\"\n",
    "\n",
    "# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq\n",
    "sample.names <- sapply(str_extract(basename(fnFs), \"T16S\\\\d\\\\d\\\\d\"), `[`, 1) ### change the pattern to the naming convention you use for the samples, in this case \"WUF\" followed by 3 digits\n",
    "\n",
    "# Generate filtered file path\n",
    "filtFs <- file.path(\"/home/amhorst/amplicon_seq/filtered/\", paste0(sample.names, \"_F_filt.fastq.gz\")) ### change to the empty folder that will hold the filtered reads for this lane (lane 1 in this case)\n",
    "filtRs <- file.path(\"/home/amhorst/amplicon_seq/filtered/\", paste0(sample.names, \"_R_filt.fastq.gz\")) ### change to the empty folder that will hold the filtered reads for this lane (lane 1 in this case)\n",
    "names(filtFs) <- sample.names\n",
    "names(filtRs) <- sample.names\n",
    "\n",
    "\n",
    "out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen = c(240,160),  #### change truncLen based on your quality scores\n",
    "                     maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE, \n",
    "                     compress=TRUE, multithread=TRUE, verbose = TRUE)\n",
    "\n",
    "out\n",
    "\n",
    "errF <- learnErrors(filtFs, multithread=TRUE, randomize = T)\n",
    "errR <- learnErrors(filtRs, multithread=TRUE, randomize = T)\n",
    "\n",
    "# Write plot to file\n",
    "pdf(\"/home/amhorst/amplicon_seq/qual_plots/err1_FWD.pdf\")   \n",
    "plotErrors(errF, nominalQ=TRUE)\n",
    "dev.off()\n",
    "\n",
    "# Write plot to file\n",
    "pdf(\"/home/amhorst/amplicon_seq/qual_plots/err1_RVS.pdf\")   \n",
    "plotErrors(errR, nominalQ=TRUE)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "dadaFs <- dada(filtFs, err=errF, multithread=TRUE)\n",
    "dadaRs <- dada(filtRs, err=errR, multithread=TRUE)\n",
    "\n",
    "dadaFs[[1]]\n",
    "dadaRs[[1]]\n",
    "\n",
    "\n",
    "\n",
    "mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)\n",
    "# Inspect the merger data.frame from the first sample\n",
    "head(mergers[[1]])\n",
    "\n",
    "\n",
    "seqtab <- makeSequenceTable(mergers)\n",
    "dim(seqtab)\n",
    "table(nchar(getSequences(seqtab)))\n",
    "\n",
    "\n",
    "\n",
    "seqtab.nochim <- removeBimeraDenovo(seqtab, method=\"consensus\", multithread=TRUE, verbose=TRUE)\n",
    "dim(seqtab.nochim)\n",
    "sum(seqtab.nochim)/sum(seqtab)\n",
    "table(nchar(getSequences(seqtab.nochim)))\n",
    "\n",
    "\n",
    "getN <- function(x) sum(getUniques(x))\n",
    "track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))\n",
    "\n",
    "colnames(track) <- c(\"input\", \"filtered\", \"denoisedF\", \"denoisedR\", \"merged\", \"nonchim\")\n",
    "rownames(track) <- sample.names\n",
    "head(track)\n",
    "\n",
    "\n",
    "taxa <- assignTaxonomy(seqtab.nochim, \"/home/csantosm/databases/silva_v138.1/silva_nr99_v138.1_train_set.fa.gz\", multithread=TRUE)\n",
    "taxa <- addSpecies(taxa, \"/home/csantosm/databases/silva_v138.1/silva_species_assignment_v138.1.fa.gz\")\n",
    "\n",
    "taxa.print <- taxa # Removing sequence rownames for display only\n",
    "rownames(taxa.print) <- NULL\n",
    "head(taxa.print)\n",
    "\n",
    "otu <- t(seqtab.nochim)\n",
    "otu.tax <- taxa %>% as.data.frame() %>% mutate(OTU_ID = row.names(.))\n",
    "rownames(otu.tax) <- NULL\n",
    "\n",
    "# Write data to RDS files\n",
    "saveRDS(seqtab.nochim, \"/home/amhorst/amplicon_seq/dada_files/otu_dada1_16s.RDS\")  \n",
    "saveRDS(taxa, \"/home/amhorst/amplicon_seq/dada_files/tax_dada1_16s.RDS\")           \n",
    "saveRDS(otu, \"/home/amhorst/amplicon_seq/dada_files/otu1_16s.RDS\")                  \n",
    "saveRDS(otu.tax, \"/home/amhorst/amplicon_seq/dada_files/tax1_16s.RDS\")             \n",
    "\n",
    "# Write data to csv files\n",
    "write.csv(seqtab.nochim, \"/home/amhorst/amplicon_seq/dada_files/otu_dada1_16s.csv\")  \n",
    "write.csv(taxa, \"/home/amhorst/amplicon_seq/dada_files/tax_dada1_16s.csv\")         \n",
    "write.csv(otu, \"/home/amhorst/amplicon_seq/dada_files/otu1_16s.csv\")                 \n",
    "write.csv(otu.tax, \"/home/amhorst/amplicon_seq/dada_files/tax1_16s.csv\")   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6916060",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
